### DISCUSSION TITLE
How I show the current branch even though there is no current branch

### DISCUSSION THREAD

**Discussion by amiryal** [jj-vcs/jj#2918] [Show and tell]
`jj` has no notion (currently) of a “current branch”, but it can tell which branches are going to be pushed on `jj git push` by default. According to the documentation, these are “any branches pointing to `remote_branches(remote=<remote>)..@`”. Feeding this to `jj log -r` will show us (among other things in the default log template) the branch we are working on, but it will return an empty result if our remote branch is already up to date:

```sh
jj log -r 'remote_branches(remote="origin")..@ & branches()'
```

To see the branch we are working on regardless of when we `jj git push`, we may change the revspec to use `trunk()` or `immutable_heads()` instead of `remote_branches(…)`:

```sh
jj log -r 'trunk()..@ & branches()'
```

Now, to see just the branch name, we can add a template:

```sh
jj log -r 'trunk()..@ & branches()' -T local_branches
```

This is already quite close to clean output, but we also don’t require the graph, and notice the keyword `branches` in there (plural). Multiple branches may refer to the same commit, or they may refer to multiple different commits in our chain of work, so we may get multiple branch names per line or multiple lines of output. Our final incantation takes care to output one branch name per line, without the graph:

```sh
jj log -r 'trunk()..@ & branches()' -T 'local_branches.join("\n") ++ "\n"' --no-graph --ignore-working-copy
```

Personally, I use this as part of my shell prompt (hence the added `--ignore-working-copy` at the end). It could also potentially be incorporated somehow into `jj status`, if it makes sense to the maintainers.

**Comment by colemickens** [comment]
I think this is a time where `--ignore-working-copy` might come in handy as well?

**Reply by amiryal** [reply]
Right! Updated the OP to include it, thanks.
